refactor: 멘토링 피드백 반영하여 이미지 처리 및 로그인 확인 훅 위치 이동(#89)

yeeun-kor opened the pull request:
<!-- PR 제목 규칙:
[type]: 작업 요약 (#이슈번호)
예시: feat: 로그인 페이지 UI 구현 (#89)
-->

## #️⃣ 연관된 이슈

- Resolves #89
- Resolves #80


## 📑 구현 사항
### 1. 이미지 로딩 최적화

- `loading="lazy"` 속성 제거하고 이미지 로드 상태 관리로 전환
- `onLoad`/`onError` 이벤트를 활용한 스켈레톤 UI 표시 개선
- 참고자료 : 지훈님코드

### 2. 무한 스크롤 성능 개선

- useEffect 기반 페이지 호출을 `useInView`의 `onChange` 콜백으로 변경
- 불필요한 리렌더링 방지 및 API 호출 최적화
- 참고자료 : `onChange` 콜백  https://www.npmjs.com/package/react-intersection-observer

### 3. 사용자 인증 로직 개선

- 멘토님의 조언 반영하여, 유저 정보 호출을 App.tsx에서 중앙화하여 초기화
- 재사용하지 않는 useAuth 훅 제거 및 로직 단순화

### 4. UI 컴포넌트 개선

- Card 컴포넌트에서 불필요한 border 제거
- 카드 컨텐츠의 flex-1 추가로 레이아웃 개선


## 🌀 어려웠던 점 / 고민했던 부분
### 접근성 개선 (Dialog 컴포넌트)
<img width="2436" height="1212" alt="image" src="https://github.com/user-attachments/assets/a8691059-1444-406b-832d-169851853e45" />

- 모달창 오픈하니 위와 같은 경고 메시지가 콘솔창에 나왔다. 
>  #### Radix UI의 DialogContent 컴포넌트는 스크린 리더(Screen Reader) 사용자를 위해 반드시 DialogTitle이 필요합니다. 
- 즉, 스크린 리더는 시각 장애인들이 웹 페이지를 읽을 수 있도록 도와주는 보조 기술인데, 
-  Radix UI 컴포넌트 사용하려면 필수라는 것
- 그러나 나는 따로 Modal내부에서 Card컴포넌트를 사용하고 싶었기 때문에, DialogTitle 숨겨주어야 했다. 

<img width="858" height="392" alt="image" src="https://github.com/user-attachments/assets/4f0eb2c0-b186-4993-900e-189af77cb05d" />

### 해결완료 
<img width="1014" height="346" alt="image" src="https://github.com/user-attachments/assets/a79c09c8-99f5-4eae-a1cc-1a353566a3eb" />

<img width="2600" height="1338" alt="image" src="https://github.com/user-attachments/assets/ed3b05a5-007c-467e-9c20-7ed4e136bc52" />

- 그래서 내린 결론은 필수요소를 추가하고, 공식문서에서 제안한 VisuallyHidden 활용한 스크린 리더 대응하였다.
- 패키지: `@radix-ui/react-visually-hidden` 추가
- 참고자료: https://www.radix-ui.com/primitives/docs/components/dialog#title

## 📸 구현 이미지
### 이미지 로딩 최적화 
#### 모달창
![모달창이미지최적화](https://github.com/user-attachments/assets/15de1a49-657c-4397-9a3e-3b12808d4a84)


#### 카드영역
![카드이미지최적화](https://github.com/user-attachments/assets/ad2a3661-bbe3-44b9-bbdd-aa506cbe0925)


### 카드 컨텐츠의 타이틀 영역을 flex-1
#### 비포
<img width="1204" height="978" alt="image" src="https://github.com/user-attachments/assets/20b314af-62b9-487f-b451-0a00435fa6c8" /> 
 
- 카드 타이틀 영역이 2줄로 떨어지면서, 카드의 푸터가 붕 뜬 상황 

#### 애프터
<img width="612" height="496" alt="스크린샷 2025-12-18 오후 2 14 17" src="https://github.com/user-attachments/assets/f64d7550-5fe5-4531-bd40-9c9104b61a12" />

- 카드 컨텐츠의 flex-1  를 줌으로써 이를 해결함

## ❇️ 코드 설명
### 스크롤 성능 개선 
<img width="740" height="538" alt="image" src="https://github.com/user-attachments/assets/5a0e55ac-8e14-4b71-a483-ef562d29a80d" />

- 페이지 호출이 한번에 2개씩 호출되고 있는 상황 

```tsx
  //무한스크롤
  const { ref } = useInView({
    threshold: 0,
    rootMargin: '50px',
    onChange(inView) {
      if (inView && hasNextPage && !isFetchingNextPage) {
        fetchNextPage()
      }
    },
  })

  //inView 변할 때 마다, 다음페이지 호출
  // useEffect(() => {
  //   if (inView && hasNextPage && !isFetchingNextPage) {
  //     fetchNextPage()
  //   }
  // }, [inView, hasNextPage, isFetchingNextPage, fetchNextPage])
```


<img width="764" height="351" alt="image" src="https://github.com/user-attachments/assets/7132feda-151f-4f93-b8a8-c6f377686b3c" />

- 멘토님의 조언대로 `useEffect` 대신, `onChange`를 활용하여 이를 해결하였다. 
- 이유는 트러블슈팅을 통해서 , 블로그 정리해서 공유 드리겠습니다. 


### 사용자 훅 위치 변경 
```tsx
function App() {
  useEffect(() => {
    const initAuth = async () => {
      try {
        if (import.meta.env.VITE_NODE_ENV === 'development') {
          const userData = await getUserInformationApi()
          useAuthStore.getState().setUser(userData)
        } else {
          try {
            const userData = await getUserInformationApi()
            useAuthStore.getState().setUser(userData)
          } catch (error) {
            console.log(error, '토큰발급에러')
            return
          }
        }
      } catch (error) {
        console.log(error, '비회원입니다')
      }
    }
    initAuth()
  }, [])
  return (
    <>
      <AppRoutes />
      <ToastContainer
        position="top-right"
        autoClose={3000}
        icon={false}
        closeButton={false}
      />
    </>
  )
}
```
- 위치 변경 하였습니다. 
## 💬 리뷰 요청사항

> 리뷰어에게 피드백을 받고 싶은 지점이 있다면 작성해주세요.

1. 카드 이미지 스켈레톤으로 대체할때, 높이를 `h-auto`로 맞추었는데, 이게 바람직한 사이즈일까요? 실제로 어떤 이미지파일이 들어올지 모르겠는데,,, 조언을 듣고싶습니다.

yeeun-kor committed:
build: yeeun-repo 동기화
- 현재 조직 repo의 develop push가 된다면?
- yeeun repo로 동일한 내용으로 동기화 되도록 cicd 워크플로우 추가함

yeeun-kor committed:
refactor: 카드 이미지 lazy로딩 수정
-  loading="lazy" 삭제
- 이미지 개선 속도 증가

yeeun-kor committed:
refactor: 유저 호출 로직 변경
- 유저 호출시 App.tsx에서 진행

yeeun-kor committed:
refactor: 인피니티스크롤 페이지 api요청 로직 수정
- onChange 를 통하여 페이지 필요할 때만 호출하기
- useEffect 는 삭제

yeeun-kor committed:
fix: shadcn Dialog컴포넌트 사용시 필수 요소 수정
- DialogContent 내부에 DialogTitle 있어야함
- 카드 컴포넌트만 보이고 싶어 DialogTitle은 필요없는 상황
- 공식문서에서 권장하는 VisuallyHidden 설치하여이를 해결함

yeeun-kor committed:
refactor: 강의 카드 UI 레이아웃 및 이미지 로딩 개선

- Card 컴포넌트에서 border 제거 및 flex-1 추가
- Accordion 트리거에서 justify-between 제거
- LectureCard 배지 및 북마크 버튼 절대 위치 조정
- 이미지 로딩 상태 관리 추가로 스켈레톤 UI 개선
- 모바일 아코디언 width 속성 조정
- LectureReviewCard 이미지 로딩 상태 관리 추가

yeeun-kor committed:
refactor: api경로 수정
- 북마크 경로수정 (기존 v1/ -> /v1/)

yeeun-kor commented on .github/workflows/deploy-to-yeeun-repo.yml:
오호 이건 제꺼 브랜치랑 미리 CICD연동한거였는데, 
현재 #88  이슈를 통해 새로 만들 예정입니다

agrade1 commented on src/components/lecture/LectureCard.tsx:
```diff
@@ -100,16 +101,15 @@ export default function LectureCard(lecture: Lecture) {
-        ) : (
-          <Skeleton></Skeleton>
-        )}
+        {/* IMG 로딩중이면 스켈레톤  */}
+        {!imgLoaded && <Skeleton className="h-[210px] w-[390px]" />}
+        <img
+          src={thumbnail_img_url}
+          alt={title}
+          onLoad={() => setImgLoaded(true)}
+          onError={() => setImgLoaded(true)}
+          className="h-auto w-full object-cover"
+        />
```
이미지 높이가 auto로 설정되어 있기 때문에,
이미지가 로드되기 전에는 브라우저가 이미지의 실제 높이를 알 수 없어서
height가 0으로 렌더링됩니다.

이후 이미지가 로드되면 실제 높이가 계산되면서
리플로우 이슈가 발생할 수 있어 성능적으로도 좋지 않습니다.

 aspect-ratio를 먼저 지정해두고, object-cover로 맞추거나 
디자인 시안상의 고정 height값을 반응형별로 지정해두는 방식이 더 안정적인 UI를 만들 수 있을 것 같습니다.

agrade1 replied:
[리플로우와 리페인트 참고 블로그](https://mong-blog.tistory.com/entry/%EB%A6%AC%ED%94%8C%EB%A1%9C%EC%9A%B0-%EB%A6%AC%ED%8E%98%EC%9D%B8%ED%8A%B8%EC%99%80-%EB%B8%8C%EB%9D%BC%EC%9A%B0%EC%A0%80-%EB%A0%8C%EB%8D%94%EB%A7%81-%EC%95%8C%EC%95%84%EB%B3%B4%EA%B8%B0#google_vignette)

yeeun-kor replied:
오호 지금 바로 수정해서 언급드리겠습니다

yeeun-kor committed:
fix: 지훈님 이미지 속성 반영
- aspect-auto  로 비율은 자동으로 맞추어놓고
- 높이값만 210으로 고정시켜두었습니다.
- 모달컴포넌트의 전체너비를 full삭제하여 그냥 일반 너비를 384픽셀로 고정하였습니다.

yeeun-kor reviewed (commented):
- 모달창의 전체너비 full제거 
- w-sm로 고정값 주었습니다 ! 
- 반응형으로 넘어가면 드롭다운으로 바뀌어서 카드 너비는 괜찮은 것 같습니다. 

<img width="1464" height="960" alt="스크린샷 2025-12-19 오후 4 10 46" src="https://github.com/user-attachments/assets/1d3d8c29-1b1b-4606-b588-ff70aea95db9" />


- aspect-auto 비율 맞추었습니다. 

<img width="1054" height="960" alt="스크린샷 2025-12-19 오후 4 10 18" src="https://github.com/user-attachments/assets/33a0836d-97f3-4b14-a03a-cf059d3c5c8a" />
<img width="1464" height="960" alt="스크린샷 2025-12-19 오후 4 10 43" src="https://github.com/user-attachments/assets/a4754826-97bc-488b-a5c8-101c8a19c6db" />
<img width="674" height="960" alt="스크린샷 2025-12-19 오후 4 10 56" src="https://github.com/user-attachments/assets/05fa94c0-4289-486a-9a62-3059f9d48e05" />

<img width="1464" height="960" alt="스크린샷 2025-12-19 오후 4 10 43" src="https://github.com/user-attachments/assets/9c8fd950-2cfe-48ab-93b6-a89b4727ee1f" />
